stripe/src/main/java/com/stripe/android/CustomerSession.java

```diff
@@ -0,0 +1,299 @@
+import com.stripe.android.exception.StripeException;
+import com.stripe.android.model.Customer;
+
+import java.lang.ref.WeakReference;
+import java.util.Calendar;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Represents a logged-in session of a single Customer.
+ */
+public class CustomerSession implements Parcelable {
+
+    private @NonNull Customer mCustomer;
+    private @NonNull EphemeralKey mEphemeralKey;
+    private @NonNull EphemeralKeyProvider mKeyProvider;
+
+    private @NonNull Handler mUiThreadHandler;
+
+    private @Nullable Calendar mProxyNowCalendar;
+    private @Nullable StripeApiProxy mStripeApiProxy;
+
+    // A queue of Runnables for doing customer updates
+    private final BlockingQueue<Runnable> mNetworkQueue = new LinkedBlockingQueue<>();
+
+    private @NonNull ThreadPoolExecutor mThreadPoolExecutor;
+
+    private static final int CUSTOMER_RETRIEVED = 7;
+
+    // The maximum number of active threads we support
+    private static final int THREAD_POOL_SIZE = 3;
+    // Sets the amount of time an idle thread waits before terminating
+    private static final int KEEP_ALIVE_TIME = 2;
+    // Sets the Time Unit to seconds
+    private static final TimeUnit KEEP_ALIVE_TIME_UNIT = TimeUnit.SECONDS;
+
+    private CustomerSession(Parcel in) {
+        ClassLoader keyProviderLoader =
+                PaymentConfiguration.getInstance().getEphemeralKeyProviderClassLoader();
+        if (keyProviderLoader == null) {
+            throw new IllegalStateException("Cannot create CustomerSession objects without " +
+                    "a KeyProvider with proper ClassLoader");
+        }
+
+        // Initialize Threads and Handlers
+        initializeMainThreadHandler();
+        initializeThreadPoolExecutor();
+
+        mKeyProvider = in.readParcelable(keyProviderLoader);
+
+        // The Ephemeral Key class is part of this package, so we use this classloader.
+        mEphemeralKey = in.readParcelable(getClass().getClassLoader());
+
+        Customer customer = Customer.fromString(in.readString());
+        mCustomer = customer == null ? Customer.getEmptyCustomer() : customer;
+    }
+
+    /**
+     * Create a CustomerSession with the provided {@link EphemeralKeyProvider}.
+     *
+     * @param keyProvider an {@link EphemeralKeyProvider} used to get
+     * {@link EphemeralKey EphemeralKeys} as needed
+     */
+    public CustomerSession(@NonNull EphemeralKeyProvider keyProvider) {
+        this(keyProvider, null, null);
+    }
+
+    @VisibleForTesting
+    CustomerSession(
+            @NonNull EphemeralKeyProvider keyProvider,
+            @Nullable StripeApiProxy stripeApiProxy,
+            @Nullable Calendar proxyNowCalendar) {
+        mKeyProvider = keyProvider;
+        mEphemeralKey = EphemeralKey.getEmptyKey();
+        mCustomer = Customer.getEmptyCustomer();
```

Comment:
Not if we need this to be `Parcelable` -- though going down to the less-performant `Serializable` interface would make this easier. But if we're making this into a Singleton any way, then there's no need to ever pass it between Activities, which obviates the need to stuff it in an Intent, which means it doesn't matter if its parcelable or serializable.